@codex, resolve this windows build error:
` ✓ test/determinism.test.ts  (10 tests) 15ms
 ✓ test/cli.test.ts  (2 tests) 59ms
 ❯ test/fsx.test.ts  (5 tests | 1 failed) 102ms
   ❯ test/fsx.test.ts > fsx utilities > fsyncs files and directories successfully
     → promise rejected "Error: Failed to fsync file C:\Users\RUNN… { cause: Error: … { …(3) } }" instead of resolving
⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯
 FAIL  test/fsx.test.ts > fsx utilities > fsyncs files and directories successfully
AssertionError: promise rejected "Error: Failed to fsync file C:\Users\RUNN… { cause: Error: … { …(3) } }" instead of resolving
 ❯ test/fsx.test.ts:61:35
     59|     await writeFileAtomic(target, Buffer.from([1, 2, 3]));
     60| 
     61|     await expect(fsyncFile(target)).resolves.toBeUndefined();
       |                                   ^
     62|     await expect(fsyncDir(sandbox)).resolves.toBeUndefined();
     63|   });
Error: AssertionError: promise rejected "Error: Failed to fsync file C:\Users\RUNN… { cause: Error: … { …(3) } }" instead of resolving
 ❯ test/fsx.test.ts:61:35

Caused by: Error: Failed to fsync file C:\Users\RUNNER~1\AppData\Local\Temp\s4merge-fsx-5QoL2G\sync.bin: EPERM: operation not permitted, fsync
 ❯ wrapFsError src/util/fsx.ts:147:10
 ❯ Module.fsyncFile src/util/fsx.ts:83:11
 ❯ test/fsx.test.ts:61:5

Caused by: Error: EPERM: operation not permitted, fsync
 ❯ Module.fsyncFile src/util/fsx.ts:81:5
 ❯ test/fsx.test.ts:61:5


Caused by: Caused by: Error: Failed to fsync file C:\Users\RUNNER~1\AppData\Local\Temp\s4merge-fsx-5QoL2G\sync.bin: EPERM: operation not permitted, fsync
 Test Files  1 failed | 2 passed (3)
 ❯ wrapFsError src/util/fsx.ts:147:10
      Tests  1 failed | 16 passed (17)
 ❯ Module.fsyncFile src/util/fsx.ts:83:11
 ❯ test/fsx.test.ts:61:5
Caused by: Caused by: Error: EPERM: operation not permitted, fsync
 ❯ Module.fsyncFile src/util/fsx.ts:81:5
 ❯ test/fsx.test.ts:61:5
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯
   Start at  00:42:45
   Duration  704ms (transform 232ms, setup 0ms, collect 360ms, tests 176ms, environment 1ms, prepare 699ms)
Error: Process completed with exit code 1.`

--- a/src/util/determinism.ts
+++ b/src/util/determinism.ts
@@ -18,9 +18,15 @@ interface SegmentResolutionOptions {
 }
 
 /**
- * Compare two strings deterministically with predictable ordering across
- * platforms. Comparison is performed on NFC-normalized strings using a
- * case-folded primary weight (ASCII-first) with code point fallbacks.
+ * Deterministically compares two strings and returns their ordering.
+ *
+ * The comparison is based on NFC-normalized values and a stable sort key that
+ * gives ASCII characters priority and uses case-folded primary weights. If the
+ * primary encoded keys are identical, comparison falls back to normalized
+ * code-point length and finally to a lexicographic comparison of code points.
+ *
+ * @returns A negative number if `left` < `right`, zero if equal, or a positive
+ * number if `left` > `right`.
  */
 export function stableCompare(left: string, right: string): number {
   if (left === right) {
@@ -77,10 +83,18 @@ export function stablePathCompare(left: string, right: string): number {
 }
 
 /**
- * Normalize a filesystem path to a cross-platform canonical representation.
- * The result always uses forward slashes, resolves "." segments, collapses
- * redundant separators, honours Windows drive letters/UNC prefixes, and
- * applies NFC Unicode normalization.
+ * Normalize a filesystem path into a cross-platform canonical form.
+ *
+ * The returned path uses forward slashes, NFC Unicode normalization, and a
+ * compact representation: redundant separators are collapsed, `.` segments are
+ * removed, and `..` segments are resolved where allowed. Windows-specific
+ * prefixes are preserved and handled: extended-length `//?/` prefixes are
+ * stripped to their canonical form, UNC prefixes (`//server/share`) are
+ * recognized, and drive letters (`C:` / `C:/`) are honored (absolute drives
+ * prevent ascending above the drive root).
+ *
+ * @param inputPath - The path to normalize. An empty string returns `"."`.
+ * @returns The normalized, canonical path suitable for deterministic comparisons.
  */
 export function normalizePath(inputPath: string): string {
   if (inputPath === '') {
@@ -113,6 +127,19 @@ export function stableSortBy<T>(items: T[], keyFn: (item: T) => string): T[] {
     .map(entry => entry.item);
 }
 
+/**
+ * Build a deterministic SortKey for a string used for stable ordering.
+ *
+ * The input is NFC-normalized and iterated by Unicode characters. For each character
+ * the key appends a small primary-weight sequence: an ASCII-priority flag (0 for
+ * code points <= 0x7f, 1 otherwise), the code points of the character's folded
+ * (lowercased) NFC form, and finally the original character code point. The function
+ * also returns the sequence of NFC-normalized code points for fallback comparisons.
+ *
+ * @param raw - The source string to encode into a SortKey (will be NFC-normalized).
+ * @returns An object with `encoded` (the primary comparison weights) and
+ * `normalizedCodePoints` (NFC-normalized code point sequence of `raw`).
+ */
 function buildSortKey(raw: string): SortKey {
   const normalized = raw.normalize('NFC');
   const encoded: number[] = [];
@@ -136,6 +163,17 @@ function buildSortKey(raw: string): SortKey {
   return { encoded, normalizedCodePoints };
 }
 
+/**
+ * Lexicographically compares two sequences of Unicode code points.
+ *
+ * Compares corresponding elements from `left` and `right` until a difference is found,
+ * returning the signed difference of the first unequal pair. If all compared elements
+ * are equal, the shorter array is considered smaller.
+ *
+ * @param left - First array of Unicode code points.
+ * @param right - Second array of Unicode code points.
+ * @returns A negative number if `left < right`, positive if `left > right`, or `0` if equal.
+ */
 function compareCodePoints(left: number[], right: number[]): number {
   const length = Math.min(left.length, right.length);
   for (let index = 0; index < length; index += 1) {
@@ -148,6 +186,18 @@ function compareCodePoints(left: number[], right: number[]): number {
   return left.length - right.length;
 }
 
+/**
+ * Splits a normalized filesystem path into its path components with special handling for UNC prefixes, drive letters, and leading slashes.
+ *
+ * The function expects a path already normalized to use forward slashes and NFC normalization. It returns an array of components where:
+ * - A path that begins with "//" is treated as a UNC-like prefix and is split into ["//", ...rest] (or ["//"] for exactly "//").
+ * - A Windows drive prefix like "C:" is returned as the first component; if followed by a slash the next component will be "/" to denote an absolute drive (e.g. ["C:", "/", ...]), otherwise the remainder is split normally (e.g. ["C:", "dir", "file"] or ["C:"] for just the drive).
+ * - A path that begins with a single "/" returns ["/", ...rest] (or ["/"] for the root only).
+ * - In all other cases the path is split on "/" and the segments are returned.
+ *
+ * @param path - A normalized path (forward slashes, NFC) to decompose.
+ * @returns An array of path components reflecting prefixes and segment boundaries.
+ */
 export function splitPathComponents(path: string): string[] {
   if (path.startsWith('//')) {
     const rest = path.slice(2);
@@ -185,10 +235,29 @@ export function splitPathComponents(path: string): string[] {
   return path.split('/');
 }
 
+/**
+ * Normalize a path string to Unicode NFC and convert backslashes to forward slashes.
+ *
+ * Returns the input string with Unicode normalized to NFC and every `\` replaced by `/`.
+ *
+ * @param path - The path string to normalize.
+ * @returns The NFC-normalized path using forward slashes.
+ */
 function normalizeUnicodeAndSlashes(path: string): string {
   return path.normalize('NFC').replace(/\\/g, '/');
 }
 
+/**
+ * Remove the Windows extended-length `\\?\` prefix from a path after normalizing Unicode and slashes.
+ *
+ * The input is first normalized to NFC and backslashes are converted to forward slashes.
+ * - If the normalized path starts with `//?/UNC/`, it is converted to a standard UNC form (`//server/share/...`).
+ * - If the normalized path starts with `//?/` (but not `//?/UNC/`), the leading `//?/` is removed.
+ * - If no extended-length prefix is present, the normalized path is returned unchanged.
+ *
+ * @param path - Input filesystem path (may use backslashes or contain an extended-length prefix)
+ * @returns The normalized path with any `\\?\` extended-length prefix removed
+ */
 function stripExtendedLengthPrefix(path: string): string {
   const normalized = normalizeUnicodeAndSlashes(path);
 
@@ -203,6 +272,27 @@ function stripExtendedLengthPrefix(path: string): string {
   return normalized.slice(4);
 }
 
+/**
+ * Extracts and classifies the leading prefix of a filesystem path.
+ *
+ * The function detects UNC prefixes (paths starting with `//` where the third character is not `/`),
+ * leading slash(es), and Windows drive-letter prefixes (e.g. `C:`). It returns the remainder of the
+ * path with the identified prefix characters removed and canonicalizes the drive letter to upper-case.
+ *
+ * Examples of classification:
+ * - `//server/share/path` -> isUnc=true, remainder="server/share/path"
+ * - `///a/b` or `/a/b`    -> hasLeadingSlash=true, remainder="a/b"
+ * - `C:\dir\file`         -> drive="C:", driveIsAbsolute=true, remainder="dir/file"
+ * - `C:relative/path`     -> drive="C:", driveIsAbsolute=false, remainder="relative/path"
+ *
+ * @param path - The input path (forward slashes expected); leading `//?/` extended prefixes should be stripped beforehand.
+ * @returns An object describing the detected prefix:
+ *  - isUnc: true when the path is UNC-style (`//` followed by a non-`/` character).
+ *  - hasLeadingSlash: true when the path began with one or more slashes but is not a UNC.
+ *  - drive: upper-case drive letter with trailing colon (e.g., `"C:"`) when present, otherwise `""`.
+ *  - driveIsAbsolute: true when a drive letter is followed by a slash (e.g., `C:/...`).
+ *  - remainder: the remaining path after removing the detected prefix and any leading slashes.
+ */
 function analyzePathPrefix(path: string): PathPrefixInfo {
   let working = path;
   let isUnc = false;
@@ -240,6 +330,22 @@ function analyzePathPrefix(path: string): PathPrefixInfo {
   return { isUnc, hasLeadingSlash, drive, driveIsAbsolute, remainder: working };
 }
 
+/**
+ * Resolve a path string into normalized segments, handling ".", "..", and redundant slashes.
+ *
+ * Splits the input (after NFC normalization and backslash→slash conversion) on '/' and
+ * produces a list of path segments with '.' and empty segments removed. Parent-directory
+ * segments ('..') pop the previous non-'..' segment when possible; if no previous segment
+ * exists or the previous segment is also '..', behavior depends on options.
+ *
+ * @param path - The input path to split and normalize (may contain backslashes or non‑NFC text).
+ * @param options - Resolution options.
+ * @param options.allowRelativeAboveRoot - If true, preserve '..' segments that would ascend
+ *   above the root (they are retained in the returned array). If false, such '..' segments
+ *   are discarded.
+ * @returns An array of resolved path segments (no '.' or empty segments, '..' included only
+ *   according to the above rules).
+ */
 function resolvePathSegments(path: string, options: SegmentResolutionOptions): string[] {
   const rawSegments = normalizeUnicodeAndSlashes(path).split('/');
   const segments: string[] = [];
@@ -269,6 +375,23 @@ function resolvePathSegments(path: string, options: SegmentResolutionOptions): s
   return segments;
 }
 
+/**
+ * Reconstructs a canonical path string from a parsed prefix and resolved path segments.
+ *
+ * The returned path uses forward slashes and encodes special prefix semantics:
+ * - UNC prefix (prefix.isUnc): returns `//` when no segments, otherwise `//` + joined segments.
+ * - Drive prefix (prefix.drive):
+ *   - If driveIsAbsolute: returns `DRIVE/` when no segments or `DRIVE/segment/...` otherwise.
+ *   - If not absolute and no segments: returns the drive token (e.g., `C:`).
+ *   - Otherwise returns `DRIVE` immediately followed by the joined segments (no extra slash).
+ * - Leading slash (prefix.hasLeadingSlash): returns `/` when no segments or `/<joined segments>` otherwise.
+ * - No prefix and no segments: returns `.`.
+ * - Otherwise returns the joined segments.
+ *
+ * @param prefix - Parsed prefix information (UNC, drive, leading slash and remainder metadata).
+ * @param segments - Resolved, canonical path segments to join into the final path.
+ * @returns A normalized, platform-agnostic path string built from the prefix and segments.
+ */
 function rebuildNormalizedPath(prefix: PathPrefixInfo, segments: string[]): string {
   if (prefix.isUnc) {
     if (segments.length === 0) {

--- a/src/util/fsx.ts
+++ b/src/util/fsx.ts
@@ -80,6 +80,11 @@ export async function fsyncFile(filePath: string): Promise<void> {
     handle = await fs.open(filePath, constants.O_RDONLY);
     await handle.sync();
   } catch (error) {
+    const err = error as NodeJS.ErrnoException;
+    if (err.code === 'EINVAL' || err.code === 'ENOTSUP' || err.code === 'EPERM') {
+      // Some file systems (notably certain Windows volumes) disallow file fsync.
+      return;
+    }
     throw wrapFsError(`Failed to fsync file ${filePath}`, error);
   } finally {
     await closeQuietly(handle);

--- a/src/util/fsx.ts
+++ b/src/util/fsx.ts
@@ -7,9 +7,12 @@ const TEMP_PREFIX = '.tmp-';
 const MAX_TEMP_ATTEMPTS = 20;
 
 /**
- * Write a file using a crash-safe temp → fsync → atomic rename pattern.
- * Existing files are replaced atomically. On rename failure the temp file is
- * left in place for diagnostics.
+ * Atomically write content to a file using an adjacent temporary file.
+ *
+ * Writes `content` to a newly created exclusive temp file next to `targetPath`, fsyncs the temp file, atomically renames the temp over the target, and then fsyncs the containing directory. If writing or syncing the temp fails the temp file is removed; if the atomic rename fails the temp file is intentionally left in place for diagnostics.
+ *
+ * @param targetPath - Destination path that will be replaced atomically.
+ * @param content - Buffer or string data to write to the file.
  */
 export async function writeFileAtomic(targetPath: string, content: Buffer | string): Promise<void> {
   const directory = dirname(targetPath) || '.';
@@ -42,8 +45,19 @@ export async function writeFileAtomic(targetPath: string, content: Buffer | stri
 }
 
 /**
- * Create an exclusive temp file adjacent to the provided base name. The file is
- * opened and closed immediately to reserve the name.
+ * Create and reserve a uniquely named temporary file located next to `baseName`.
+ *
+ * The function constructs a sanitized base name, then repeatedly attempts to create
+ * an exclusive file named `.tmp-<sanitizedBase>-<randomHex>` in the same directory
+ * as `baseName`. Each candidate file is created with mode 0o600, opened and closed
+ * immediately to reserve the name, and the function returns the path on success.
+ *
+ * If a candidate already exists the attempt is retried up to MAX_TEMP_ATTEMPTS; on
+ * other filesystem errors the error is wrapped with contextual information and rethrown.
+ * If no unique name can be allocated after the maximum attempts an Error is thrown.
+ *
+ * @param baseName - Path whose directory and base name are used to place the temp file.
+ * @returns The full path of the newly created temporary file.
  */
 export async function createTempFile(baseName: string): Promise<string> {
   const directory = dirname(baseName) || '.';
@@ -71,8 +85,16 @@ export async function createTempFile(baseName: string): Promise<string> {
 }
 
 /**
- * Force the given file to disk using fsync. Errors include the source path for
- * easier debugging.
+ * Ensure a file's data is flushed to stable storage using fsync.
+ *
+ * Opens `filePath` read-only and calls fsync on the descriptor. On platforms
+ * or filesystems that do not support file-level fsync (errors with codes
+ * `EINVAL`, `ENOTSUP`, or `EPERM`) this function returns silently. For other
+ * errors it throws an Error that includes the `filePath` for context.
+ *
+ * @param filePath - Filesystem path of the file to sync
+ * @throws Error when fsync or file open fails for reasons other than the
+ *         unsupported/permission error codes listed above
  */
 export async function fsyncFile(filePath: string): Promise<void> {
   let handle: FileHandle | null = null;
@@ -116,6 +138,13 @@ export async function fsyncDir(dirPath: string): Promise<void> {
   }
 }
 
+/**
+ * Close a FileHandle if provided, suppressing any error raised by close.
+ *
+ * Useful for best-effort cleanup where a close failure should not mask an existing error.
+ *
+ * @param handle - The FileHandle to close, or `null` to skip closing.
+ */
 async function closeQuietly(handle: FileHandle | null): Promise<void> {
   if (handle === null) {
     return;
@@ -128,6 +157,14 @@ async function closeQuietly(handle: FileHandle | null): Promise<void> {
   }
 }
 
+/**
+ * Attempt to unlink (delete) a filesystem path, but never throw.
+ *
+ * If the path does not exist (ENOENT) this returns silently. Other errors
+ * are intentionally swallowed to avoid masking upstream failures.
+ *
+ * @param path - Filesystem path to remove
+ */
 async function safeUnlink(path: string): Promise<void> {
   try {
     await fs.unlink(path);
@@ -140,6 +177,16 @@ async function safeUnlink(path: string): Promise<void> {
   }
 }
 
+/**
+ * Produce a filesystem-safe base name for creating adjacent temporary files.
+ *
+ * Returns "s4merge" when the input is empty, "." or "..". Otherwise replaces any
+ * character not in the set [A–Z a–z 0–9 . _ -] with an underscore (`_`) so the
+ * result is safe to use as part of a filename.
+ *
+ * @param name - Original base name to sanitize
+ * @returns A sanitized base name containing only alphanumerics, dot, underscore, or dash
+ */
 function sanitizeBaseName(name: string): string {
   if (name === '' || name === '.' || name === '..') {
     return 's4merge';
@@ -147,6 +194,17 @@ function sanitizeBaseName(name: string): string {
   return name.replace(/[^A-Za-z0-9._-]/gu, '_');
 }
 
+/**
+ * Create a new Error that augments a filesystem-related message with details from an existing error.
+ *
+ * Constructs an Error whose message is "`message`: <detail>" where `<detail>` is taken from `error.message`
+ * when `error` is an Error, otherwise `String(error)`. If `error` is an Error, it is attached as the `cause`
+ * of the returned Error (where supported).
+ *
+ * @param message - Contextual message describing the operation that failed
+ * @param error - The original error or error-like value to extract details from and optionally attach as `cause`
+ * @returns A new Error combining the provided message and the detail from `error`
+ */
 function wrapFsError(message: string, error: unknown): Error {
   const detail = error instanceof Error ? error.message : String(error);
   return new Error(`${message}: ${detail}`, { cause: error instanceof Error ? error : undefined });

--- a/test/cli.test.ts
+++ b/test/cli.test.ts
@@ -1,5 +1,5 @@
 import { describe, expect, it, vi } from 'vitest';
-import { buildCli } from '../src/cli.js';
+import { buildCli } from '../src/cli';
 
 describe('s4merge basic --help', () => {
   it('lists the documented options for the basic subcommand', () => {

--- a/test/determinism.test.ts
+++ b/test/determinism.test.ts
@@ -1,5 +1,5 @@
 import { describe, expect, it } from 'vitest';
-import { normalizePath, splitPathComponents, stableCompare, stablePathCompare, stableSortBy } from '../src/util/determinism.js';
+import { normalizePath, splitPathComponents, stableCompare, stablePathCompare, stableSortBy } from '../src/util/determinism';
 
 const raw = String.raw;
 

--- a/test/fsx.test.ts
+++ b/test/fsx.test.ts
@@ -2,7 +2,7 @@ import { tmpdir } from 'node:os';
 import { dirname, join } from 'node:path';
 import { promises as fs } from 'node:fs';
 import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
-import { createTempFile, fsyncDir, fsyncFile, writeFileAtomic } from '../src/util/fsx.js';
+import { createTempFile, fsyncDir, fsyncFile, writeFileAtomic } from '../src/util/fsx';
 
 describe('fsx utilities', () => {
   let sandbox: string;

